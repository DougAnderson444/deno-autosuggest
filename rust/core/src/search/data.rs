use std::borrow::Cow;
use std::default::Default;
use super::Scores;
use crate::lexis::{Text, tokenize_record};


#[derive(Clone, Debug)]
pub struct Record {
    id: usize,
    text: Text<Vec<char>>,
}


impl Record {
    pub fn new<T: IntoIterator<Item=char>>(id: usize, source: T) -> Record {
        let source: Vec<char> = source.into_iter().collect();
        Record {
            id,
            text: tokenize_record(&source).to_owned(),
        }
    }

    pub fn to_hit<'a>(&'a self) -> Hit<Cow<'a, [char]>> {
        Hit::new(self.id, self.text.to_cow())
    }
}


#[derive(Debug, Clone)]
pub struct Hit<T: AsRef<[char]>> {
    pub id:      usize,
    pub text:    Text<T>,
    pub scores:  Scores,
}


impl<T: AsRef<[char]>> Hit<T> {
    pub fn new(id: usize, text: Text<T>) -> Hit<T> {
        Hit { id, text, scores: Default::default() }
    }
}


#[derive(Debug, Clone)]
pub struct SearchResult {
    pub id:          usize,
    pub highlighted: Vec<char>,
}
