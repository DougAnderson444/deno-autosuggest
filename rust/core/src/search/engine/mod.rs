mod score;

use std::cmp::Ordering;
use crate::lexis::Text;
use super::{Record, Hit};
use score::score;
pub use score::Scores;


pub struct Engine<'a> {
    query: Text<&'a [char]>,
    hits:  Vec<Hit<'a>>,
    limit: usize,
}


impl<'a> Engine<'a> {
    pub fn new(query: Text<&'a [char]>, limit: usize) -> Self {
        Self {
            query,
            hits: Vec::with_capacity(limit * 2),
            limit,
        }
    }

    pub fn push(&mut self, record: &'a Record) {
        let mut hit = Hit::from_record(record);
        score(&self.query, &mut hit);
        if self.matches(&hit) {
            self.hits.push(hit);
        }
        if self.hits.len() >= self.limit * 2 {
            self.sort_and_truncate();
        }
    }

    pub fn push_many<I: IntoIterator<Item=&'a Record>>(&mut self, records: I) {
        for record in records {
            self.push(record);
        }
    }

    pub fn matches(&self, hit: &Hit<'a>) -> bool {
        let matches = &hit.scores.matches;
        if self.query.is_empty() { return true; }
        if matches.len() == 0 { return false; }
        if matches.len() == 1 && self.query.words.len() > 1 {
            let word_match = &matches[0];
            let unfinished = !word_match.fin;
            let first_half = (word_match.query.len * 2) < word_match.record.len;
            if unfinished && first_half { return false; }
        }
        true
    }

    pub fn sort_and_truncate(&mut self) {
        self.hits.sort_by(|hit1, hit2| {
            hit1.scores.linear.iter()
                .zip(hit2.scores.linear.iter())
                .map(|(s1, s2)| s2.cmp(s1))
                .find(|&ord| ord != Ordering::Equal)
                .unwrap_or(Ordering::Equal)
        });
        self.hits.truncate(self.limit);
    }

    pub fn hits<'b>(&'b self) -> &'b [Hit<'a>] {
        &self.hits
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::tokenize_query;
    use super::{Record, Engine};

    fn check(name: &str, queries: &[&str]) {
        let records = [
            Record::new(10, "brown plush bear"),
            Record::new(20, "metal detector"),
            Record::new(30, "yellow metal mailbox"),
        ];
        for (i, query) in queries.iter().enumerate() {
            let query = tokenize_query(query);
            let mut finder = Engine::new(query.to_ref(), 10);
            finder.push_many(&records);
            finder.sort_and_truncate();
            assert_debug_snapshot!(format!("{}-{}", name, i), finder.hits);
        }
    }

    #[test]
    fn search_empty() {
        check("empty", &[""]);
    }

    #[test]
    fn search_equal() {
        check("equal", &["yelow metall maiblox"]);
    }

    #[test]
    fn search_partial() {
        check("partial", &[
            "brown plush bear",
            "metal detector",
            "yellow metal mailbox",
        ]);
    }

    #[test]
    fn search_intersection() {
        check("intersection", &[
            "red wooden mailbox",
            "red wooden mail",
        ]);
    }

    #[test]
    fn search_min_match() {
        check("min_match", &[
            "wooden mai",
            "wooden mail",
        ]);
    }

    #[test]
    fn search_transpositions() {
        check("transpositions", &[
            "metal mailbox",
            "mailbox metal",
        ]);
    }
}
