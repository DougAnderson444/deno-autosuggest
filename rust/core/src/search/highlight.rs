use crate::lexis::{Text, WordMatch};
use super::{Hit, Scores};

pub fn highlight<T: AsRef<[char]>>(hits: &[Hit<T>], left: &[char], right: &[char]) -> Vec<Vec<char>> {
    hits.iter()
        .map(|hit| highlight_one(hit, left, right))
        .collect()
}

fn highlight_one<T: AsRef<[char]>>(hit: &Hit<T>, left: &[char], right: &[char]) -> Vec<char> {
    let Hit { 
        scores: Scores { matches, .. }, 
        text: Text { words, source }, 
        ..
    } = hit;

    let capacity   = source.as_ref().len() + words.len() * (left.len() + right.len() + 1);
    let mut result = Vec::with_capacity(capacity);

    for (i, w) in words.iter().enumerate() {
        match matches.iter().find(|m| m.record.pos == i) {
            Some(WordMatch { record: m, .. }) => {
                let match_start = w.slice.0 + m.slice.0;
                let match_end   = w.slice.0 + m.slice.1;
                result.extend(&w.source.as_ref()[.. match_start]);
                result.extend(left);
                result.extend(&w.source.as_ref()[match_start .. match_end]);
                result.extend(right);
                result.extend(&w.source.as_ref()[match_end .. ]);
            },
            None => {
                result.extend(w.source.as_ref());
            },
        }
        result.push(' ');
    }
    result.pop();

    result
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::{Text, Chars};
    use super::super::{Hit, search};
    use super::highlight;
    use std::borrow::Cow;

    fn chars(s: &str) -> Vec<char> {
        s.chars().collect()
    }

    fn record<'a>(chars: &'a [char]) -> Text<Cow<'a, [char]>> {
        Text::new_cow(Cow::Borrowed(&chars[..]))
            .split(&Chars::Whitespaces)
            .strip(&Chars::NotAlphaNum)
            .lower()
    }

    fn query<'a>(s: &'a [char]) -> Text<Cow<'a, [char]>> {
        record(s).fin(false)
    }

    #[test]
    fn test_highlight() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("Metal detector");
        let cr3 = chars("yellow Metal Mailbox");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("metall mail");
        let q  = query(&cq);

        search(&q, &mut hits);

        let left:  Vec<char> = "[".chars().collect();
        let right: Vec<char> = "]".chars().collect();
        let highlighted: Vec<String> = highlight(&hits, &left, &right)
            .iter()
            .map(|h| h.iter().collect())
            .collect();

        assert_debug_snapshot!(highlighted);
    }

    #[test]
    fn test_highlight_stripped() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("Metal detector");
        let cr3 = chars("yellow 'Metal' -Mailbox-");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("metall mail");
        let q  = query(&cq);

        search(&q, &mut hits);

        let left:  Vec<char> = "{{".chars().collect();
        let right: Vec<char> = "}}".chars().collect();
        let highlighted: Vec<String> = highlight(&hits, &left, &right)
            .iter()
            .map(|h| h.iter().collect())
            .collect();

        assert_debug_snapshot!(highlighted);
    }
}