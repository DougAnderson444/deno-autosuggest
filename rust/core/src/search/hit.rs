use std::default::Default;
use crate::tokenization::TextRef;
use crate::matching::WordMatch;
use crate::store::Record;
use super::score::Scores;


#[derive(Debug)]
pub struct Hit<'a> {
    pub id:      usize,
    pub title:   TextRef<'a>,
    pub rating:  usize,
    pub rmatches: Vec<WordMatch>,
    pub qmatches: Vec<WordMatch>,
    pub scores:  Scores,
}


impl<'a> Hit<'a> {
    pub fn from_record(record: &'a Record) -> Hit<'a> {
        Hit {
            id:       record.id,
            title:    record.title.to_ref(),
            rating:   record.rating,
            scores:   Default::default(),
            rmatches: Vec::new(),
            qmatches: Vec::new(),
        }
    }
}
