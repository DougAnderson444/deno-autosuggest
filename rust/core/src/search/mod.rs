mod score;
mod highlight;

use std::borrow::Cow;
use std::cmp::Ordering;
use crate::lexis::{Text, Chars};

pub use score::{score, Scores};
pub use highlight::highlight;


#[derive(Clone, Debug)]
pub struct Record {
    id: usize,
    text: Text<Vec<char>>,
}


impl Record {
    pub fn new(id: usize, source: &[char]) -> Record {
        Record {
            id,
            text: tokenize_record(source).to_owned(),
        }
    }

    pub fn to_hit<'a>(&'a self) -> Hit<Cow<'a, [char]>> {
        Hit {
            id: self.id,
            text: self.text.to_cow(),
            scores: Default::default(),
        }
    }
}


#[derive(Debug, Clone)]
pub struct Hit<T: AsRef<[char]>> {
    pub id:      usize,
    pub text:    Text<T>,
    pub scores:  Scores,
}


impl<T: AsRef<[char]>> Hit<T> {
    pub fn new(id: usize, text: Text<T>) -> Hit<T> {
        Hit { id, text, scores: Default::default() }
    }
}


pub fn search<T: AsRef<[char]>, U: AsRef<[char]>>(query: &Text<T>, hits: &mut Vec<Hit<U>>) {
    score(query, hits);
    filter(query, hits);
    sort(hits);
}


pub fn filter<T: AsRef<[char]>, U: AsRef<[char]>>(query: &Text<T>, hits: &mut Vec<Hit<U>>) {
    if query.is_empty() {
        return;
    }
    hits.retain(|hit| {
        let matches = &hit.scores.matches;
        if matches.len() == 0 { return false; }
        if matches.len() == 1 && query.words.len() > 1 {
            let word_match = &matches[0];
            let unfinished = !word_match.fin;
            let first_half = (word_match.query.len * 2) < word_match.record.len;
            if unfinished && first_half { return false; }
        }
        true
    });
}


pub fn sort<T: AsRef<[char]>>(hits: &mut Vec<Hit<T>>) {
    hits.sort_by(|hit1, hit2| {
        let s1 = &hit1.scores;
        let s2 = &hit2.scores;
        if s1.matches.len() != s2.matches.len() { return s1.matches.len().cmp(&s2.matches.len()).reverse(); }
        if s1.typos  != s2.typos  { return s1.typos.cmp(&s2.typos); }
        if s1.trans  != s2.trans  { return s1.trans.cmp(&s2.trans); }
        if s1.fin    != s2.fin    { return s1.fin.cmp(&s2.fin).reverse(); }
        if s1.offset != s2.offset { return s1.offset.cmp(&s2.offset); }
        Ordering::Equal
    });
}


pub fn tokenize_record<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .split(&[Chars::Whitespaces, Chars::Control])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


pub fn tokenize_query<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::{Text, Chars};
    use super::{Hit, search};
    use std::borrow::Cow;

    fn chars(s: &str) -> Vec<char> {
        s.chars().collect()
    }

    fn record<'a>(chars: &'a [char]) -> Text<Cow<'a, [char]>> {
        Text::new_cow(Cow::Borrowed(&chars[..]))
            .split(&Chars::Whitespaces)
    }

    fn query<'a>(s: &'a [char]) -> Text<Cow<'a, [char]>> {
        record(s).fin(false)
    }

    #[test]
    fn search_empty() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }

    #[test]
    fn search_equal() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("yelow metall maiblox");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }

    #[test]
    fn search_partial() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();
        let mut hits3 = hits1.clone();

        let cq1 = chars("metall mailbox");
        let cq2 = chars("metall mail");
        let cq3 = chars("met");
        let q1  = query(&cq1);
        let q2  = query(&cq2);
        let q3  = query(&cq3);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);
        search(&q3, &mut hits3);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
        assert_debug_snapshot!(hits3);
    }

    #[test]
    fn search_intersection() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();

        let cq1 = chars("red wooden mailbox");
        let cq2 = chars("red wooden mail");
        let q1  = query(&cq1);
        let q2  = query(&cq2);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
    }

    #[test]
    fn search_min_match() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();

        let cq1 = chars("wooden mai");
        let cq2 = chars("wooden mail");
        let q1  = query(&cq1);
        let q2  = query(&cq2);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
    }

    #[test]
    fn search_transpositions() {
        let cr1 = chars("metal mailbox");
        let cr2 = chars("mailbox metal");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
        ];

        let cq = chars("mailbox met");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }
}
