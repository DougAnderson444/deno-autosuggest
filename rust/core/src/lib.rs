mod utils;

mod damlev;
mod lexis;
mod search;

use std::cell::RefCell;
use search::{Record, Hit, tokenize_query};


thread_local! {
    static SEPARATORS: RefCell<(Vec<char>, Vec<char>)> = RefCell::new((Vec::new(), Vec::new()));
    static RECORDS:    RefCell<Vec<Record>>            = RefCell::new(Vec::new());
    static HIGHLIGHTS: RefCell<Vec<Vec<char>>>         = RefCell::new(Vec::new());
}


pub fn get_highlights() -> String {
    HIGHLIGHTS.with(|cell| {
        let buffer = &*cell.borrow();
        let mut result = String::new();
        for highlighted in buffer {
            for ch in highlighted {
                result.push(*ch);
            }
            result.push('\0');
        }
        result
    })
}


pub fn highlight_using(l: &str, r: &str) {
    SEPARATORS.with(|cell| {
        let buffer = &mut *cell.borrow_mut();
        buffer.0 = l.chars().collect();
        buffer.1 = r.chars().collect();
    })
}


pub fn set_records(ids: &[usize], sources: String) {
    RECORDS.with(|cell| {
        let records = &mut *cell.borrow_mut();
        records.clear();
        for (source, id) in sources.split('\0').zip(ids.iter()) {
            let source: Vec<char> = source.chars().collect();
            records.push(Record::new(*id, &source));
        }
        if records.len() != ids.len() {
            panic!("Got less texts than expected");
        }
    });
}


pub fn run_search(query: String) -> Vec<usize> {
    RECORDS.with(|cell| {
        let query: Vec<char> = query.chars().collect();
        let query = tokenize_query(&query);

        let records  = &*cell.borrow();
        let mut hits = records.iter()
            .map(|r| r.to_hit())
            .collect();

        search::search(&query, &mut hits);
        save_highlights(&hits);

        hits.iter()
            .map(|h| h.id)
            .collect()
    })
}


fn save_highlights<T: AsRef<[char]>>(hits: &[Hit<T>]) {
    SEPARATORS.with(|cell_sep| {
    HIGHLIGHTS.with(|cell_hl| {
        let sep = &*cell_sep.borrow_mut();
        let hl  = &mut *cell_hl.borrow_mut();
        *hl = search::highlight(&hits, &sep.0, &sep.1);
    }); });
}
