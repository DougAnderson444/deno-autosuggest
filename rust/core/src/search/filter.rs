use crate::tokenization::TextRef;
use crate::search::Hit;

pub fn hit_matches(query: &TextRef, hit: &Hit) -> bool {
    if query.is_empty() { return true; }
    if hit.rmatches.len() == 0 { return false; }
    if hit.rmatches.len() == 1 && hit.qmatches.len() == 1 && query.words.len() > 1 {
        let rmatch     = &hit.rmatches[0];
        let qmatch     = &hit.qmatches[0];
        let first_half = (qmatch.word_len() * 2) < rmatch.word_len();
        if !rmatch.fin && first_half { return false; }
    }
    true
}
