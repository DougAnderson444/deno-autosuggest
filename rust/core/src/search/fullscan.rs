use crate::store::Record;
use crate::search::Hit;


pub struct FullScan<'a, Src: Iterator<Item=&'a Record>> {
    source: Src,
}


impl<'a, Src: Iterator<Item=&'a Record>> FullScan<'a, Src> {
    pub fn new(source: Src) -> Self {
        Self { source }
    }
}


impl<'a, Src: Iterator<Item=&'a Record>> Iterator for FullScan<'a, Src> {
    type Item = Hit<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.source
            .next()
            .map(Hit::from_record)
    }
}
