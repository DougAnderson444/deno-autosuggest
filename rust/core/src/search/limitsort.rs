use std::cmp::Ordering;
use crate::search::Hit;


pub struct LimitSort<'a, Src: Iterator<Item=Hit<'a>>> {
    source: Src,
    buffer: Vec<Hit<'a>>,
    limit:  usize,
    done:   bool,
}


impl<'a, Src: Iterator<Item=Hit<'a>>> LimitSort<'a, Src> {
    pub fn new(source: Src, limit: usize) -> Self {
        Self {
            source,
            buffer: Vec::with_capacity(limit * 2),
            limit,
            done: false,
        }
    }

    fn sort_and_truncate(&mut self) {
        self.buffer.sort_by(|hit1, hit2| {
            hit1.scores.linear.iter()
                .zip(hit2.scores.linear.iter())
                .map(|(s1, s2)| s2.cmp(s1))
                .find(|&ord| ord != Ordering::Equal)
                .unwrap_or(Ordering::Equal)
        });
        self.buffer.truncate(self.limit);
    }
}


impl<'a, Src: Iterator<Item=Hit<'a>>> Iterator for LimitSort<'a, Src> {
    type Item = Hit<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if !self.done {
            while let Some(hit) = self.source.next() {
                self.buffer.push(hit);
                if self.buffer.len() >= self.limit * 2 {
                    self.sort_and_truncate();
                }
            }
            self.sort_and_truncate();
            self.buffer.reverse();
            self.done = true;
        }

        self.buffer.pop()
    }
}
