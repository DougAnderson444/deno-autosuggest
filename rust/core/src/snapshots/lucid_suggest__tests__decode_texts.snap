---
source: src/lib.rs
expression: hits
---
[
    [
        'F',
        'o',
        'o',
        ',',
        ' ',
        'b',
        'a',
        'r',
    ],
    [
        'B',
        'a',
        'z',
        '-',
        '2',
    ],
    [
        'Q',
        'u',
        'x',
        '!',
    ],
]
